src/DeviceInterface.cpp

```diff
@@ -0,0 +1,133 @@
+#include "JITModule.h"
+#include "StmtCompiler.h"
+#include "Target.h"
+
+#include "runtime/HalideRuntime.h"
+#include "runtime/HalideRuntimeCuda.h"
+#include "runtime/HalideRuntimeOpenCL.h"
+#include "runtime/HalideRuntimeOpenGL.h"
+
+using namespace Halide;
+using namespace Halide::Internal;
+
+namespace {
+
+template <typename fn_type>
+bool lookup_runtime_routine(const char *name, const Target &target, fn_type &result) {
+    StmtCompiler stmt_compiler(target);
+    std::vector<JITModule> runtime(JITSharedRuntime::get(stmt_compiler.get_codegen(), target));
```

Comment:
Checking if the dev field is non-zero before loading a runtime is likely a worthwhile performance improvement. However it is not correct to assume that if there isn't a runtime defined in JITModule's shared_runtimes global, then one doesn't need to be loaded. This is because one could allocate a buffer on the GPU, then call JITSharedRuntime::release_all and then call halide_device_free on the buffer. With the current code, this will make a new runtime to be able to call the halide_device_free in the just created runtime, which will end up calling the GPU specific free routine in the previously existing runtime (which is still in memory due to ref counting, but isn't available in the globals that hold the current runtime because those were cleared to release the ref count...). Convoluted to be sure, but I think it works. Andrew pointed out that we run the risk of freeing code that we are currently executing, though I think even that is handled now as the ref count will be released for the old runtime while running in the new one...

Note also that the routines to get an interface (e.g. halide_cuda_device_interface) can legitimately be called before any runtime has been loaded for other reasons to get a device interface to use for halide_device_malloc.

I'm entirely willing to admit this is confusing. I'd love to have something simpler...